yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr

```diff
@@ -0,0 +1,364 @@
+    }
+
+    #[aztec(public)]
+    fn set_admin(
+        new_admin: AztecAddress,
+    ) {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.admin.read() == context.msg_sender(), "caller is not admin");
+        storage.admin.write(new_admin.address);
+    }
+
+    #[aztec(public)]
+    fn set_minter(
+        minter: AztecAddress,
+        approve: Field,
+    ) {
+        assert((approve == 1) | (approve == 0), "not providing boolean");
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.admin.read() == context.msg_sender(), "caller is not admin");
+        storage.minters.at(minter.address).write(approve as Field);
+    }
+
+    #[aztec(public)]
+    fn mint_pub(
+        to: AztecAddress,
+        amount: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.minters.at(context.msg_sender()).read() == 1, "caller is not minter");
+        let amount = SafeU120::new(amount);
+        let new_balance = SafeU120::new(storage.public_balances.at(to.address).read()).add(amount);
+        let supply = SafeU120::new(storage.total_supply.read()).add(amount);
+
+        storage.public_balances.at(to.address).write(new_balance.value as Field);
+        storage.total_supply.write(supply.value as Field);
+        1
+    }
+
+    #[aztec(public)]
+    fn mint_priv(
+        amount: Field,
+        secret_hash: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.minters.at(context.msg_sender()).read() == 1, "caller is not minter");
+        let pending_shields = storage.pending_shields;
+        let mut note = TransparentNote::new(amount, secret_hash);
+        let supply = SafeU120::new(storage.total_supply.read()).add(SafeU120::new(amount));
+
+        storage.total_supply.write(supply.value as Field);
+        pending_shields.insert_from_public(&mut note);
+        1
+    }
+
+    #[aztec(public)]
+    fn shield(
+        from: AztecAddress,
+        amount: Field,
+        secret_hash: Field,
+        nonce: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+
+        if (from.address != context.msg_sender()) {
+            // The redeem is only spendable once, so we need to ensure that you cannot insert multiple shields from the same message.
+            let selector = compute_selector("shield((Field),Field,Field,Field)");
+            let message_field = compute_message_hash([context.msg_sender(), selector, from.address, amount, secret_hash, nonce]);
+            assert(AccountContract::at(from.address).is_valid(Option::none(), Option::some(context), message_field) == 0xe86ab4ff, "invalid call");
+            context.push_new_nullifier(message_field, 0);
+        } else {
+            assert(nonce == 0, "invalid nonce");
```

Comment:
Mainly because I did not like a value that could be whatever when not used, so making it 0 made it pretty clear as well if doing a direct transfer or the transfer from.